### ISSUE TITLE
Rule Change: `no-constant-binary-expression` should detect type mismatches in primitive wrapper comparisons

### ISSUE THREAD

**Issue by kimtaejin3** [eslint/eslint#20964]
### What rule do you want to change?

no-constant-binary-expression

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
Boolean(x) === 1;            // boolean vs number → always false
Boolean(x) === "true";        // boolean vs string → always false
Boolean(x) === 0;             // boolean vs number → always false

Number(x) === "string";       // number vs string → always false
Number(x) === "1";            // number vs string → always false
Number(x) === true;           // BASELINE — already flagged

String(x) === 1;              // string vs number → always false
String(x) === false;          // string vs boolean → always false
String(x) === null;           // BASELINE — already flagged
```

### What does the rule currently do for this code?

The rule currently only flags comparisons in these cases:

1. One side is `null`/`undefined` and the other has constant non-nullish type
   → `Number(x) === null` is flagged 
2. One side is a static boolean (`true`, `false`, `!constant`) and the other is
   not coerceable to that boolean
   → `Number(x) === true` is flagged 
3. One side is a newly-constructed object
   → `Boolean(x) === {}` is flagged 

however, type-level mismatches between two non-null primitives are not detected

- `Boolean(x) === 1`         → passes (boolean vs number)
- `Number(x) === "string"`   → passes (number vs string)
- `String(x) === 1`           → passes (string vs number)

### What will the rule do after it's changed?

The rule should also flag comparisons between a primitive wrapper call and a literal of a mismatched primitive type, since `===` between values of different types is always `false` (and `!==` always `true`).

### Participation

- [x] I am willing to submit a pull request to implement this change.

### AI acknowledgment

- [ ] I did not use AI to generate this issue report.
- [x] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by DMartens** [comment]
I would be -1 on this change as the intent of the rule is to find common problems and I do not think a developer would accidently compare a casted value to a different type.
For example one of the the motivating example for the rule was `value === []` as developers coming from other languages may think JavaScript supports this.
TypeScript has a sophisticated type checker which catches this and many others.

**Comment by kimtaejin3** [comment]
Thanks for the feedback @DMartens I understand — the rule is meant to catch common developer mistakes, and a type mismatch between a casted value and a different primitive is unlikely to be written by accident.
Agreed that type-level checks like this fit better at the TypeScript layer.
